File: app/src/main/java/de/jerleo/model/Bill.java
package de.jerleo.model;

import android.util.JsonReader;
import android.util.JsonWriter;

import androidx.annotation.NonNull;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import de.jerleo.android.DateHelper;
import de.jerleo.database.Constants;
import de.jerleo.database.Database;

public class Bill implements Comparable<Bill> {

    private Home home;
    private long id;
    private String description;
    private Calendar begin;
    private List<BillItem> items, removeItems;
    private List<Meter> meters;

    public static List<Bill> importDatabase() {

        return Database.getBillList();
    }

    public static List<Bill> importFrom(JsonReader reader) throws IOException {

        List<Bill> bills = new ArrayList<>();

        reader.beginArray();
        while (reader.hasNext())
            bills.add(Bill.importSingleFrom(reader));
        reader.endArray();

        return bills;
    }

    private static Bill importSingleFrom(JsonReader reader) throws IOException {

        Bill bill = new Bill();

        reader.beginObject();
        while (reader.hasNext()) {
            String name = reader.nextName();

            switch (name) {
                case Constants.DESCRIPTION:
                    bill.setDescription(reader.nextString());
                    break;
                case Constants.BEGIN:
                    bill.setBegin(DateHelper.getDatabaseDate(reader.nextString()));
                    break;
                case Constants.METERS:
                    bill.setItems(BillItem.importFrom(reader));
                    break;
                default:
                    reader.skipValue();
                    break;
            }
        }
        reader.endObject();

        return bill;
    }

    @Override
    public int compareTo(@NonNull Bill other) {

        final int compareBegin = this.begin.compareTo(other.begin);
        final int compareDescription = this.description.compareTo(other.description);
        if (compareBegin == 0)
            return compareDescription;
        else
            return compareBegin;
    }

    public void delete() {

        for (final BillItem item : items)
            item.delete();
        items.clear();

        Database.delete(this);
    }

    public void exportTo(JsonWriter writer) throws IOException {

        writer.beginObject();

        writer.name(Constants.DESCRIPTION).value(description);
        writer.name(Constants.BEGIN).value(DateHelper.getDatabaseDate(begin));

        if (items.size() > 0) {
            writer.name(Constants.METERS);
            writer.beginArray();
            for (final BillItem item : items)
                item.exportTo(writer);
            writer.endArray();
        }

        writer.endObject();
    }

    public Calendar getBegin() {

        return (Calendar) begin.clone();
    }

    public String getDescription() {

        return description;
    }

    public Calendar getEnd() {

        Calendar end = (Calendar) begin.clone();
        end.add(Calendar.YEAR, 1);
        end.add(Calendar.DAY_OF_MONTH, -1);
        return end;
    }

    public Home getHome() {

        return home;
    }

    public long getId() {

        return id;
    }

    public List<Meter> getMeters() {

        if (meters == null)
            meters = new ArrayList<>();

        meters.clear();

        for (final BillItem item : items)
            meters.add(item.getMeter());

        return meters;
    }

    public float getTotalCosts() {

        float costs = 0;
        for (final Meter meter : getMeters())
            costs += meter.getCosts(this);
        return costs;
    }

    public float getTotalFees() {

        float fees = 0;
        for (final Meter meter : getMeters())
            fees += meter.getFees(this);
        return fees;
    }

    public float getTotalPayments() {

        float payments = 0;
        for (final Meter meter : getMeters())
            payments += meter.getPayments(this);
        return payments;
    }

    public void persist() {

        if (id == 0)
            id = Database.insert(this);
        else
            Database.update(this);

        if (removeItems != null) {
            for (BillItem item : removeItems)
                Database.delete(item);
            removeItems = null;
        }

        for (BillItem item : items)
            item.persist();
    }

    public void remove(Meter meter) {

        for (BillItem item : items) {
            if (item.getMeter().equals(meter)) {
                item.delete();
                items.remove(item);
                return;
            }
        }
    }

    public void setBegin(Calendar begin) {

        this.begin = begin;
    }

    public void setDescription(String description) {

        this.description = description;
    }

    public void setHome(Home home) {

        this.home = home;
    }

    public void setId(long id) {

        this.id = id;
    }

    public void setItems(List<BillItem> newItems) {

        if (items != null) {
            removeItems = new ArrayList<>();

            for (final BillItem item : items)
                if (!newItems.contains(item))
                    removeItems.add(item);
        }
        for (final BillItem item : newItems)
            item.setBill(this);
        items = newItems;
    }
}

File: app/src/main/java/de/jerleo/model/Bill.kt
package de.jerleo.model

import android.util.JsonReader
import android.util.JsonWriter
import de.jerleo.android.DateHelper
import de.jerleo.database.Constants
import de.jerleo.database.Database
import de.jerleo.database.Database.Companion.delete
import de.jerleo.database.Database.Companion.insert
import de.jerleo.database.Database.Companion.update
import java.io.IOException
import java.time.LocalDate

class Bill : Comparable<Bill> {

    var id: Long = 0L
    var name: String = ""

    lateinit var home: Home
    lateinit var dateTo: LocalDate
        private set

    var dateFrom: LocalDate = DateHelper.today
        set(value) {
            field = value
            dateTo = value.plusYears(1).minusDays(1)
        }

    var items: MutableList<Item> = ArrayList()
        set(value) {
            removed = field
            field.clear()
            field.addAll(value)
            field.forEach { it.bill = this }
        }

    private var removed: MutableList<Item> = ArrayList()
        set(value) {
            field.clear()
            field.addAll(value)
        }

    override fun compareTo(other: Bill): Int {
        val compareBegin = dateFrom.compareTo(other.dateFrom)
        val compareDescription = name.compareTo(other.name)
        return if (compareBegin == 0) compareDescription else compareBegin
    }

    fun usage(year:Int) = items.sumOf { it.usage(year) }
    fun payments(year: Int) = items.sumOf { it.payments(year) }
    fun fees(year: Int) = items.sumOf { it.fees(year) }
    fun costs(year: Int) = items.sumOf { it.costs(year) }

    fun delete() {
        items.forEach { it.delete() }
        items.clear()
        delete(this)
    }

    fun persist() {
        if (id == 0L) id = insert(this) else update(this)
        removed.forEach { delete(it) }
        removed.clear()
        items.forEach { it.persist() }
    }

    private fun minYear(): Int {
        val min = items.mapNotNull {
            it.meter!!.firstReading()?.date
        }.minOfOrNull { it } ?: return 0
        val from = LocalDate.of(min.year, dateFrom.month, dateFrom.dayOfMonth)
        return min.year - if (min < from) 1 else 0
    }

    private fun maxYear(): Int {
        val max = items.mapNotNull {
            it.meter!!.lastReading()?.date
        }.maxOfOrNull { it } ?: return 0
        val from = LocalDate.of(max.year, dateFrom.month, dateFrom.dayOfMonth)
        return max.year - if (max < from) 1 else 0
    }

    fun months(year: Int): MutableList<LocalDate> = DateHelper.months(
        dateFrom.withYear(year),
        DateHelper.min(
            DateHelper.today.withDayOfMonth(1),
            dateFrom.withYear(year + 1).minusDays(1)
        )
    )

    fun years(): MutableList<Int> = DateHelper.years(minYear(), maxYear())
    fun contains(meter: Meter) = items.find { it.meter == meter } != null
    fun remove(meter: Meter) {
        items.filter { it.meter == meter }.forEach {
            it.delete()
            items.remove(it)
        }
    }

    @Throws(IOException::class)
    fun exportTo(writer: JsonWriter) {
        writer.beginObject()
        writer.name(Constants.DESCRIPTION).value(name)
        writer.name(Constants.BEGIN).value(DateHelper.format(dateFrom))
        if (items.isNotEmpty()) {
            writer.name(Constants.METERS)
            writer.beginArray()
            items.forEach { it.exportTo(writer) }
            writer.endArray()
        }
        writer.endObject()
    }

    companion object {

        val import by lazy { Database.bills }

        @Throws(IOException::class)
        fun import(reader: JsonReader): MutableList<Bill> {
            val bills: MutableList<Bill> = ArrayList()
            reader.beginArray()
            while (reader.hasNext()) bills.add(importSingle(reader))
            reader.endArray()
            return bills
        }

        @Throws(IOException::class)
        private fun importSingle(reader: JsonReader): Bill {
            val bill = Bill()
            reader.beginObject()
            while (reader.hasNext()) {
                when (reader.nextName()) {
                    Constants.DESCRIPTION -> bill.name = reader.nextString()
                    Constants.BEGIN -> bill.dateFrom = DateHelper.parse(reader.nextString())
                    Constants.METERS -> bill.items = Item.import(reader)
                    else -> reader.skipValue()
                }
            }
            reader.endObject()
            return bill
        }
    }
}